File: src/Typecheck.hs
module Typecheck where

import Control.Monad
import Control.Monad.State
import Control.Monad.Except
import qualified Data.Map as Map
import qualified Data.Set as Set
import Control.Applicative ((<$>))

import Ast
import Type
import Ident

type Infer a = ExceptT String (State Int) a

runInfer :: Infer (Subst, Type) -> Either String Type
runInfer m = snd <$> evalState (runExceptT m) 0

type TypeEnv = Map.Map ValIdent Scheme

lookupEnv :: TypeEnv -> ValIdent -> (Subst, Type)
lookupEnv env x = case Map.lookup x env of
  Nothing -> throwError $ "Unbound variable: " ++ show x
  Just s -> do
    t <- instantiate s
    return (nullSubst, t)

type Subst = Map.Map String Type

nullSubst :: Subst
nullSubst = Map.empty

compose :: Subst -> Subst -> Subst
s `compose` t = Map.map (subst s) t `Map.union` s

class Substitutable a where
  subst :: Subst -> a -> a
  freeTVars :: a -> Set.Set String

instance Substitutable Type where
  subst s t = case t of
    TVar a -> Map.findWithDefault t a s
    TApp t ts -> TApp t (map (subst s) ts)
    TArrow t u -> TArrow (subst s t) (subst s u)
    TImplicit n t u ->
      let s' = n `Map.delete` s in
      TImplicit n (subst s' t) (subst s' u)
    TNamed n t ->
      let s' = n `Map.delete` s in
      TNamed n (subst s' t)
    TRec n t ->
      let s' = n `Map.delete` s in
      TRec n (subst s' t)
    TDual t -> TDual $ subst s t
    TOffer s os ->
      let os' = map (\(n, i, r) -> (n, subst s i, subst s r)) os in
      TOffer s os'

  freeTVars t = case t of
    TRec tv t -> tv `Set.delete` freeTVars t
    TImplicit _ t u -> freeTVars t `Set.union` freeTVars u
    TNamed _ t -> freeTVars t
    TVar tv -> Set.singleton tv
    TApp _ ts -> Set.unions $ map freeTVars ts
    TArrow t u -> freeTVars t `Set.union` freeTVars u
    TDual t -> freeTVars t
    TOffer s os -> Set.unions $ map (\(_, i, r) -> freeTVars i `Set.union` freeTVars r) os

instance Substitutable Scheme where -- )FIXME? Constraints treated wrong(?)
  subst s (Forall ts t) = Forall (map substConstraint ts) (subst s' t)
    where s' = foldr Map.delete s ts
          substConstraint (s, Nothing) = (s, Nothing)
          substConstraint (s, Just (CEqual t)) = (s, Just $ CEqual $ subst s' t)
          substConstraint (s, Just (CSubtype t)) = (s, Just $ CSubtype $ subst s' t)

  freeTVars (Forall [] t) = freeTVars t
  freeTVars (Forall ((v, Nothing):vs) t) = freeTVars (Forall vs t) `Set.delete` v
  freeTVars (Forall ((v, Just (CEqual c)):vs) t) = (freeTVars (Forall vs t) `Set.delete` v) `Set.union` freeTVars t
  freeTVars (Forall ((v, Just (CSubtype c)):vs) t) = (freeTVars (Forall vs t) `Set.delete` v) `Set.union` freeTvars t

letters :: [String]
letters = [1..] >>= flip replicateM ['a'..'z']

fresh :: Infer Type
fresh = do
  u <- get
  lift $ modify (+1)
  return $ TVar (letters !! u)

occurs :: String -> Type -> Bool
occurs tv t = tv `Set.member` freeTVars t

-- FIXME? Check for infinite types
bind :: String -> Type -> Infer Subst
bind a t | t == TVar a = return nullSubst
         | otherwise   = return $ Map.singleton a t

-- TODO: Unify type-level identifiers?
unify :: Type -> Type -> Infer Subst
unify (TVar a) t = bind a t
unify t (TVar a) = bind a t
unify t u = case (t, u) of
  (TRec tv t, TRec tv', t') -> do
    s1 <- bind tv (TVar tv')
    s2 <- unify (subst s1 t) (subst s1 t')
    return $ s2 `compose` s1
  (TImplicit _ t u, TImplicit _ t' u') -> do
    s1 <- unify t t'
    s2 <- unify (subst s1 u) (subst s1 u')
  (TNamed _ t, TNamed _ t') -> unify t t'
  (TApp a ts, TApp b ts') ->
    if a == b
      then foldr (uncurry compose) nullSubst (zip ts ts')
      else throwError "Unification failed on TApp"
  (TArrow l r, TArrow l' r') -> do
    s1 <- unify l l'
    s2 <- unify (subst s1 r) (subst s1 r')
    return $ s2 `compose` s1
  (TDual t, TDual t') -> unify t t'
  (TOffer _ os, TOffer _ os') -> do
    let unifyOffers [] [] = nullSubst
        unifyOffers (_, i, r):os (_, i', r'):os' = do
          s1 <- unify i i'
          s2 <- unify (subst s1 r) (subst s1 r')
          s3 <- unifyOffers os os'
          return $ s3 `compose` s2 `compose` s1 in
    unifyOffers os os'

instantiate :: Scheme -> Infer Type
instantiate = undefined

generalize :: TypeEnv -> Type -> Scheme
generalize = undefined

class Inferable a where
  infer :: TypeEnv -> a -> Infer (Subst, Type)

-- FIXME: Allow letrec/mutual recursion
instance forall a. Inferable (Expr a) where
  infer env e = case e of
    ELiteral (LBool _)    -> return (nullSubst, tBool)
    ELiteral (LInteger _) -> return (nullSubst, tInteger)
    ELiteral (LDecimal _) -> return (nullSubst, tFloat)
    ELiteral (LString _)  -> return (nullSubst, tString)

    EIdent v -> lookupEnv env x

    ELet id x e -> do
      (s1, t1) <- infer env x
      let env'  = subst s1 env
          t'    = generalize env' t1
      (s2, t2) <- infer (Map.insert id t' env) e
      return (s1 `compose` s2, t2)

    ESession bs -> do
      env <- prepareEnv bs
      let inferBranches [] = fresh >>= \tv -> return (nullSubst, TOffer tv [])
          inferBranches (b:bs) = case b of
            TInit es -> do
              (s1, t1) <- infer env es
              (s2, t2) <- inferBranches (subst s1 bs)
              -- FIXME? I think this (may) require unifying c & t2
              return $ case t1 of
                TBranch x os -> TBranch x $ map (\(l, t, c) -> (l, t, t2)) os
                TChoose x os -> TChoose x $ map (\(l, t, c) -> (l, t, t2)) os
                _ -> t2
            TBranch msg args tgt e ->

    ECond pred t f ->

    ELambda id e ->

    EApply f x ->

    EBranch msg args tgt e ->

    EChoose tgt msg ->

    ESpawn v ->

    ESequence s ->

instance forall a. Inferable (Top a) where
  infer env e = case e of
    TModule m ts -> return $ infer env ts -- FIXME: Handle module scoping?

    TRequire m -> return (nullSubst, tInt) -- FIXME: Import scope?

    TLet id x -> do
      placeHolder <- lookupEnv env id
      (s, t) <- infer env x
      s' <- unify placeHolder (subst s t)
      let env' = subst s' env
          t'  = generalize env' t
      return (s', t')

    TInit e -> infer env e

    TBranch msg args tgt e ->

instance Inferable x => Inferable [x] where
  infer env [] = fresh >>= \tv -> (nullSubst, tv)
  infer env (x:xs) = do
    (s, _) <- infer env x
    (ss, t) <- infer env (subst s xs)
    return (s `compose` ss, t)

prepareEnv :: forall a. [Top a] -> Infer TypeEnv
prepareEnv [] = return Map.empty
prepareEnv (t:ts) = case t of
  TLet id _ -> do
    tyId <- fresh
    return $ Map.insert id tyId (prepareEnv ts)
  _ -> prepareEnv ts -- FIXME: Should imports, etc. be handled here?

File: src/typecheck.ml
open Ast
open Type

module Scope = Map.Make(struct
    type t = identifier
    let compare = compare
  end)

module Subst = Map.Make(struct
    type t = Type.t
    let compare = compare
  end)

type env =
  { scope : Type.t Scope.t
  ; subst : Type.t Subst.t
  ; state : int
  }

let fresh env =
  ({ env with state = env.state + 1 }, TVar env.state)

let rec apply env = function
  | TVar _ as tvar ->
    if Subst.mem tvar env.subst
        then apply env (Subst.find tvar env.subst)
        else tvar
  | TApp (f, xs) -> TApp (f, List.map (apply env) xs)
  | TArrow (a, b) -> TArrow (List.map (apply env) a, apply env b)
  | TImplicit (n, t, u) -> TImplicit (n, apply env t, apply env u)
  | TNamed (n, t) -> TNamed (n, apply env t)
  | TRec (r, t) -> TRec (r, apply env t)
  | TOffer (w, offers) -> TOffer (w, List.map (fun (n, p, c) -> (n, apply env p, apply env c)) offers)
  | TChoose (w, offers) -> TChoose (w, List.map (fun (n, p, c) -> (n, apply env p, apply env c)) offers)
  | TQuant (q, l, t) -> TQuant (q, l, apply env t)

let rec contains(t2) = function
  | TApp (_, xs) -> List.exists (contains(t2)) xs
  | TArrow (a, b) -> List.exists (contains(t2)) a || contains(t2) b
  | TImplicit (_, a, b) -> contains(t2) a || contains(t2) b
  | TNamed (_, t) -> contains(t2) t
  | TRec (_, t) -> contains(t2) t
  | TOffer (_, offers) | TChoose (_, offers) -> List.exists (fun (_, p, c) -> contains(t2) p || contains(t2) c) offers
  | TQuant (_, _, t) -> contains(t2) t
  | t1 -> t1 = t2

let rec unify env t1 t2 =
  let t1 = apply env t1 and t2 = apply env t2 in
  match t1, t2 with
  | _, _ when t1 = t2 -> env
  | TVar _, _ when not (contains(t2) t1) -> { env with subst = Subst.add t1 t2 env.subst }
  | _, TVar _ when not (contains(t1) t2) -> { env with subst = Subst.add t2 t1 env.subst }
  | TApp (f, xs), TApp (g, ys) when f = g -> List.fold_left2 unify env xs ys
  | TArrow (args1, r1), TArrow (args2, r2) -> unify (List.fold_left2 unify env args1 args2) r1 r2
  | TImplicit (n1, t1, c1), TImplicit (n2, t2, c2) when n1 = n2 -> unify (unify env t1 t2) c1 c2
  | TNamed (n1, t1), TNamed (n2, t2) when n1 = n2 -> unify env t1 t2
  | TRec (r1, t1), TRec (r2, t2) -> unify (unify env (TVar r1) (TVar r2)) t1 t2
  | TQuant (q1, l1, t1), TQuant (q2, l2, t2) when l1 = l2 -> unify (unify env (TVar q1) (TVar q2)) t1 t2
  | TOffer (w1, o1), TOffer (w2, o2) | TChoose (w1, o1), TChoose (w2, o2) when w1 = w2 ->
    begin
      let module Offers = Set.Make(String)
      and name_of_offer (n, _, _) = n in
      let names = Offers.(elements @@ union (of_list @@ List.map name_of_offer o1) (of_list @@ List.map name_of_offer o2)) in
      let rec loop env = function
        | [] -> env
        | name::names ->
          begin
            let matcher = List.find (fun o -> name_of_offer o = name) in
            match (matcher o1, matcher o2) with
              | (_, t1, c1), (_, t2, c2) ->
                let env = unify env t1 t2 in
                let env = unify env c1 c2 in
                loop env names
              | exception Not_found -> loop env names
          end in
      loop env names
    end
  | _, _ -> failwith @@ "Could not unify type " ^ string_of_type t1 ^ " with type " ^ string_of_type t2
